fid_drift_mon/update_fid_stats.py

```diff
@@ -0,0 +1,269 @@
+    if acen is None:
+        return
+
+    LOGGER.info(f"Got {len(fidprops)} rows fidprops and {len(acen)} rows acacen")
+
+    # Calculate fid statistics, returning a list of dict for each fid
+    stats = calc_fid_stats(obs, fidprops, acen)
+
+    # Write the fid statistics to the database
+    for stat in stats:
+        LOGGER.debug(f"Inserting row {stat}")
+        dbh.insert(stat, "fid_stats")
+
+
+def get_archive_file_data(obs, content, max_files=None):
+    files = asp_l1.get_files(
+        start=obs["obs_start"], stop=obs["obs_stop"], content=[content]
+    )
+
+    if len(files) == 0:
+        LOGGER.info(
+            f"No {content} files found for obsid {obs['obsid']} between "
+            f"{obs['obs_start']} and {obs['obs_stop']}"
+        )
+        return None
+
+    if max_files is not None and len(files) > max_files:
+        LOGGER.error(
+            f"Too many {content} files found for obsid {obs['obsid']} between "
+            f"{obs['obs_start']} and {obs['obs_stop']}"
+        )
+        return None
+
+    dats = [Table.read(file) for file in files]
+    LOGGER.debug(f"Got {len(dats)} {content} files")
+    if len(dats) == 1:
+        out = dats[0]
+    else:
+        out = vstack(dats, join_type="exact", metadata_conflicts="silent")
+
+    return out
+
+
+def delete_obs(dbh, obsid):
+    LOGGER.info(f"Deleting fid_stats entries for obsid {obsid}")
+    dbh.execute(f"delete from fid_stats where obsid={obsid}")
+
+
+def calc_fid_stats(obs, fidprops, acen):
+    """
+    Calculate fid statistics
+    """
+    stats = []
+    for fidpr in fidprops:
+        if fidpr["id_status"].strip() != "GOOD":
+            LOGGER.info(
+                f"WARNING: obsid {obs['obsid']} has bad fid status "
+                f"{fidpr['id_status']!r} in slot {fidpr['slot']}"
+            )
+            continue
+        stat = calc_stats_for_fidpr(obs, acen, fidpr)
+        if stat is not None:
+            stats.append(stat)
+
+    return stats
+
+
+def calc_stats_for_fidpr(obs, acen, fidpr):
+    t_int = 1.696
+    slot = fidpr["slot"]
+    ok = (acen["slot"] == slot) & (acen["alg"] == 8) & (acen["status"] == 0)
+    cen = acen[ok]
+    n_cen = len(cen)
+
+    if n_cen < 200:
+        LOGGER.info(
+            f"WARNING: obsid {obs['obsid']} has {n_cen} < 200 "
+            f"fid counts in slot {slot}, skipping"
+        )
+        return
+
+    props_samples = 1000 if n_cen >= 1000 else n_cen
+
+    ang_y_sm = cen["ang_y_sm"] * 3600
+    ang_z_sm = cen["ang_z_sm"] * 3600
+    # SIM_Z offset in arcsec (using 20.493 arcsec/mm)
+    sim_z_offset = (fidpr.meta["LSI0STT3"] + fidpr.meta["STT0STF3"]) * 20.493
+    stat = {
+        "slot": slot,
+        "id_num": fidpr["id_num"],
+        "id_string": fidpr["id_string"],
+        "tstart": CxoTime(obs["obs_start"]).secs,
+        "obsid": obs["obsid"],
+        "date_obs": obs["obs_start"],
+        "ang_y_med": np.median(ang_y_sm[0:props_samples]),
```

Comment:
This had been left with a CHANGE ME comment, as it was set to use [-1000:0].  I think maybe the first 1000 samples (unless observation is shorter) is perfectly fine for our use cases, but I actually don't know our use cases @taldcroft !  So what are thoughts about that?

Reply:
I have no idea about the driver for using the first 4 ksec rather than last 4 ksec in the comment, but I guess I had an opinion at the time. I guess it corresponds a bit more to fid acquisition.

In any case, you can just change to a slice like `[:1000]`. This works regardless of the length of the object being sliced so you don't need that `props_sample` variable. Or define `props_sample_max = 1000` and use that in the slice.

Reply:
Yeah, I was going back and forth on it... though mostly thinking that we have no strong dependence on these data either way.  Though I thought [-1000:] was actually wrong for short observations.

Reply:
(With regard to a driver, I was thinking that the CHANGE_ME was related to comparison with fidprops as I think the pipeline just uses the initial samples  as a reference for the prop files --- but I actually don't know)

Reply:
```
In [1]: a = [1, 2, 3]

In [2]: a[-1000:]
Out[2]: [1, 2, 3]
```

Reply:
For historical context when we had long observations, using the end of the observation would normally be more thermally stable. Even now the end is more stable than the beginning for many observations.

I do think that using the original `[-1000:]` will help with functional regression testing since you should be getting close to the same answers.

Reply:
Right, didn't mean python indexing wrong, I mean conceptually "this isn't the median of the last 1000 samples of the observation".  Which was just as wrong for "first 1000 samples" when it is short.   It looks like the code has decided that if it has fewer than 200 samples not to get stats, but it seems like we should at least save the number of samples in the stats.